packages/http/src/validator/validators/body.ts

```diff
@@ -1,82 +1,131 @@
-
+    const content = getContent(specs, mediaType);
     const schema = get(content, 'schema');
 
-    if (!schema) {
-      return [];
-    }
-
-    if (mediaType && typeIs.is(mediaType, 'application/x-www-form-urlencoded')) {
-      const shallowParsedTarget = shallowParseUrlEncoded(target);
-      target = deepParseUrlEncoded(target);
-
-      const encodings = get(content, 'encodings', []);
-      for (const encoding of encodings) {
-        const allowReserved = get(encoding, 'allowReserved', false);
-        const property = encoding.property;
-        const value = shallowParsedTarget[property];
-
-        if (!allowReserved && typeof value === 'string' && value.match(/[\/?#\[\]@!$&'()*+,;=]/)) {
-          return [
-            {
-              path: [prefix, property],
-              message: 'Reserved characters used in request body',
-              severity: DiagnosticSeverity.Error,
-            },
-          ];
-        }
+    return pipe(
+      schema ? Either.right(target) : Either.left([]),
+      Either.chain(partial(maybeValidateFormBody, schema!, content, mediaType)),
+      Either.chain(partial(validateBody, schema!)),
+      Either.fold(partial(applyPrefix, this.prefix), () => []),
+    );
+  }
+}
 
-        if (encoding.style) {
-          const deserializer = body.get(encoding.style);
-          if (deserializer && schema.properties) {
-            const propertySchema = schema.properties[property];
-            target[property] = deserializer.deserialize(property, target, propertySchema as JSONSchema);
-          }
-        }
-      }
-    }
+function validateBody(schema: JSONSchema, target: any) {
+  const diagnostics = validateAgainstSchema(target, schema);
+  return diagnostics.length ? Either.left(diagnostics) : Either.right(target);
+}
+
+function maybeValidateFormBody(
+  schema: JSONSchema,
+  content: IMediaTypeContent,
+  mediaType: string | undefined,
+  target: any,
+) {
+  if (mediaType && typeIs.is(mediaType, 'application/x-www-form-urlencoded')) {
+    const encodings = get(content, 'encodings', []);
+    const encodedUriParams = splitUriParams(target);
 
-    return validateAgainstSchema(target, schema).map(error =>
-      Object.assign({}, error, { path: [prefix, ...(error.path || [])] }),
+    return pipe(
+      validateAgainstReservedCharacters(encodedUriParams, encodings),
+      Either.map(decodeUriEntities),
+      Either.map(deserializeFormBody.bind(undefined, schema, encodings)),
     );
   }
 
-  private getContent(specs: IMediaTypeContent[], mediaType?: string): IMediaTypeContent | undefined {
-    if (!mediaType) {
-      return specs[0];
-    }
+  return Either.right(target);
+}
+
+function applyPrefix(prefix: string, diagnostics: IPrismDiagnostic[]): IPrismDiagnostic[] {
+  return diagnostics.map(d => ({ ...d, path: [prefix, ...(d.path || [])] }));
+}
+
+function validateAgainstReservedCharacters(encodedUriParams: Dictionary<string, string>, encodings: IHttpEncoding[]) {
```

Comment:
I think I wrote this somewhere else, but for monads I suggest to explicitly put the return type, it's going to help TypeScript a lot when doing type checking. 
```suggestion
function validateAgainstReservedCharacters(encodedUriParams: Dictionary<string, string>, encodings: IHttpEncoding[]): Either.Either<IPrismDiagnostic[], Dictionary<string, string>> {
```

Reply:
Done 👍